lib/core/src/restore.rs

```diff
@@ -0,0 +1,1164 @@
+//! This module provides functionality for restoring the swap tx IDs from onchain data
+
+use std::collections::HashMap;
+
+use anyhow::{anyhow, Result};
+use log::{error, info};
+use lwk_wollet::elements::Txid;
+use lwk_wollet::WalletTx;
+use sdk_common::bitcoin::hashes::hex::ToHex;
+
+use crate::prelude::*;
+use crate::restore::immutable::*;
+
+/// A map of all our known LWK onchain txs, indexed by tx ID. Essentially our own cache of the LWK txs.
+pub(crate) struct TxMap {
+    outgoing_tx_map: HashMap<Txid, WalletTx>,
+    incoming_tx_map: HashMap<Txid, WalletTx>,
+}
+impl TxMap {
+    pub(crate) fn from_raw_tx_map(raw_tx_map: HashMap<Txid, WalletTx>) -> Self {
+        let (outgoing_tx_map, incoming_tx_map): (HashMap<Txid, WalletTx>, HashMap<Txid, WalletTx>) =
+            raw_tx_map
+                .into_iter()
+                .partition(|(_, tx)| tx.balance.values().sum::<i64>() < 0);
+
+        Self {
+            outgoing_tx_map,
+            incoming_tx_map,
+        }
+    }
+}
+
+trait PartialSwapState {
+    /// Determine partial swap state, based on recovered chain data.
+    /// For example, it cannot distinguish between [PaymentState::Created] and [PaymentState::TimedOut],
+    /// and in some cases, between [PaymentState::Created] and [PaymentState::Failed].
+    fn get_partial_state(&self) -> PaymentState;
+}
+
+pub(crate) struct RecoveredOnchainDataSend {
+    lockup_tx_id: Option<HistoryTxId>,
+    claim_tx_id: Option<HistoryTxId>,
+    refund_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataSend {
+    fn get_partial_state(&self) -> PaymentState {
+        match &self.lockup_tx_id {
+            Some(_) => match &self.claim_tx_id {
+                Some(_) => PaymentState::Complete,
+                None => match &self.refund_tx_id {
+                    Some(refund_tx_id) => match refund_tx_id.confirmed {
+                        true => PaymentState::Failed,
+                        false => PaymentState::RefundPending,
+                    },
+                    None => PaymentState::Pending,
+                },
+            },
+            None => PaymentState::TimedOut,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataReceive {
+    lockup_tx_id: Option<HistoryTxId>,
+    claim_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataReceive {
+    fn get_partial_state(&self) -> PaymentState {
+        match (&self.lockup_tx_id, &self.claim_tx_id) {
+            (Some(_), Some(claim_tx_id)) => match claim_tx_id.confirmed {
+                true => PaymentState::Complete,
+                false => PaymentState::Pending,
+            },
+            (Some(_), None) => PaymentState::Pending,
+            // TODO How to distinguish between Failed and Created (if in both cases, no lockup or claim tx present)
+            //   See https://docs.boltz.exchange/v/api/lifecycle#reverse-submarine-swaps
+            _ => PaymentState::Failed,
```

Comment:
If no lockup or claim present then it means the payer didn't start paying (swapper didn't broadcast the lockup) from our perspective it is always Created (even on normal flow without restore).

Reply:
No lockup or claim tx can mean both Created (no payment happened yet) or Failed (`invoice.expired`, `swap.expired`, or even `transaction.failed`).

We default to Created now, but I left this comment to point out that we lose this Boltz-specific status information.

Reply:
I see. Wasn't aware that we set it to Failed in these cases. I am not sure BTW we need to set it to Failed in case of receiving but this is another story out of the scope of this issue.